### PR TITLE
[Merged by Bors] - feat: port LinearAlgebra.Matrix.BilinearForm

### PR DESCRIPTION
**Pull request by Vierkantor**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by Vierkantor** [comment]
I'll be away for a few hours, please feel free to help out!

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+In this file we use the following type variables:
+ - `M`, `M'`, ... are modules over the semiring `R`,
+ - `M₁`, `M₁'`, ... are modules over the ring `R₁`,
+ - `M₂`, `M₂'`, ... are modules over the commutative semiring `R₂`,
+ - `M₃`, `M₃'`, ... are modules over the commutative ring `R₃`,
+ - `V`, ... is a vector space over the field `K`.
+
+## Tags
+
+bilinear_form, matrix, basis
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:41]: These names are still mathlib3.

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+
+variable {n o : Type _}
+
+open BigOperators
+
+open BilinForm Finset LinearMap Matrix
+
+open Matrix
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:107:6: warning: expanding binder group (i j) -/
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:68]: Is this line produced by mathport? Similarly elsewhere.

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+/-- The map from `Matrix n n R` to bilinear forms on `n → R`.
+
+This is an auxiliary definition for the equivalence `matrix.to_bilin_form'`. -/
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:71]: I am not sure what happened here, but there is no `matrix.to_bilin_form'` (not even in mathlib3 I mean). Maybe an old doc? In any case it's probably a good idea to fix it and use the right mathlib4 name (I don't think it is worth to fix mathlib3 and forward port the change). Similarly below.

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean [hunk 4]
```diff
@@ -0,0 +1,623 @@
...
+    Matrix.toBilin'Aux (BilinForm.toMatrixAux (R₂ := R₂)
+      (fun j => stdBasis R₂ (fun _ => R₂) j 1) B₂) = B₂ := by
+  refine' ext_basis (Pi.basisFun R₂ n) fun i j => _
+  rw [Pi.basisFun_apply, Pi.basisFun_apply, Matrix.toBilin'Aux_stdBasis,
+    BilinForm.toMatrixAux_apply]
+#align to_bilin'_aux_to_matrix_aux toBilin'Aux_toMatrixAux
+
+section ToMatrix'
+
+/-! ### `to_matrix'` section
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:133]:
```suggestion
/-! ### `ToMatrix'` section
```

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean [hunk 5]
```diff
@@ -0,0 +1,623 @@
...
+    M.toBilin'.comp (Matrix.toLin' P) (Matrix.toLin' Q) = Matrix.toBilin' (Pᵀ ⬝ M ⬝ Q) :=
+  BilinForm.toMatrix'.injective
+    (by simp only [BilinForm.toMatrix'_comp, BilinForm.toMatrix'_toBilin', toMatrix'_toLin'])
+#align matrix.to_bilin'_comp Matrix.toBilin'_comp
+
+end ToMatrix'
+
+section ToMatrix
+
+/-! ### `to_matrix` section
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:285]:
```suggestion
/-! ### `ToMatrix` section
```

**Review by riccardobrasca — APPROVED** [review]
LGTM, thanks!

bors d+

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean (hunk 1 above)
Comment (Vierkantor) [reply Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:41]:
```suggestion
bilinear form, bilin form, BilinearForm, matrix, basis
```

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean (hunk 2 above)
Comment (Vierkantor) [reply Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:68]: Yes, I recall the best practice was to keep these warnings so we can debug mathport, but I'd be happy to remove them of course.

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean (hunk 3 above)
Comment (Vierkantor) [reply Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:71]:
```suggestion
This is an auxiliary definition for the equivalence `Matrix.toBilin'`. -/
```

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+    apply mul_eq_zero_of_left
+    exact stdBasis_ne R₂ (fun _ => R₂) _ _ hi' 1
+  · intros
+    have := Finset.mem_univ i
+    contradiction
+#align matrix.to_bilin'_aux_std_basis Matrix.toBilin'Aux_stdBasis
+
+/-- The linear map from bilinear forms to `Matrix n n R` given an `n`-indexed basis.
+
+This is an auxiliary definition for the equivalence `matrix.to_bilin_form'`. -/
```
Comment (Vierkantor) [inline Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:106]:
```suggestion
This is an auxiliary definition for the equivalence `Matrix.toBilin'`. -/
```

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean (hunk 2 above)
Comment (Vierkantor) [reply Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:68]: Apparently I remembered wrong: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/unsupported.3A.20ambiguous.20notation/near/338596270

```suggestion
```

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean [hunk 7]
```diff
@@ -0,0 +1,623 @@
...
+def Matrix.toBilin' : Matrix n n R₂ ≃ₗ[R₂] BilinForm R₂ (n → R₂) :=
+  BilinForm.toMatrix'.symm
+#align matrix.to_bilin' Matrix.toBilin'
+
+@[simp]
+theorem Matrix.toBilin'Aux_eq (M : Matrix n n R₂) : Matrix.toBilin'Aux M = Matrix.toBilin' M :=
+  rfl
+#align matrix.to_bilin'_aux_eq Matrix.toBilin'Aux_eq
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:107:6: warning: expanding binder group (i j) -/
```
Comment (Vierkantor) [inline Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:171]:
```suggestion
```

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean [hunk 8]
```diff
@@ -0,0 +1,623 @@
...
+#align matrix.to_bilin Matrix.toBilin
+
+@[simp]
+theorem BilinForm.toMatrix_apply (B : BilinForm R₂ M₂) (i j : n) :
+    BilinForm.toMatrix b B i j = B (b i) (b j) := by
+  rw [BilinForm.toMatrix, LinearEquiv.trans_apply, BilinForm.toMatrix'_apply, congr_apply,
+    b.equivFun_symm_stdBasis, b.equivFun_symm_stdBasis]
+#align bilin_form.to_matrix_apply BilinForm.toMatrix_apply
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:107:6: warning: expanding binder group (i j) -/
```
Comment (Vierkantor) [inline Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:313]:
```suggestion
```

**Comment by Vierkantor** [comment]
bors merge

**Pull request closed without merging**